app/vminsert/common/insert_ctx.go

```diff
@@ -60,15 +61,21 @@ func (ctx *InsertCtx) marshalMetricNameRaw(prefix []byte, labels []prompb.Label)
 }
 
 // WriteDataPoint writes (timestamp, value) with the given prefix and labels into ctx buffer.
-func (ctx *InsertCtx) WriteDataPoint(prefix []byte, labels []prompb.Label, timestamp int64, value float64) error {
+func (ctx *InsertCtx) WriteDataPoint(prefix []byte, labels []prompbmarshal.Label, timestamp int64, value float64) error {
+	if storagelimits.ExceedingLabels(labels) {
+		return nil
+	}
 	metricNameRaw := ctx.marshalMetricNameRaw(prefix, labels)
 	return ctx.addRow(metricNameRaw, timestamp, value)
 }
 
 // WriteDataPointExt writes (timestamp, value) with the given metricNameRaw and labels into ctx buffer.
 //
 // It returns metricNameRaw for the given labels if len(metricNameRaw) == 0.
-func (ctx *InsertCtx) WriteDataPointExt(metricNameRaw []byte, labels []prompb.Label, timestamp int64, value float64) ([]byte, error) {
+func (ctx *InsertCtx) WriteDataPointExt(metricNameRaw []byte, labels []prompbmarshal.Label, timestamp int64, value float64) ([]byte, error) {
```

Comment:
almost each vminsert ingestion protocol counts all samples even if they are dropped later
[here](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/37f1c76e717425ea587650c809f3906228b235e5/app/vminsert/datadogv1/request_handler.go#L46) we increment rowsLen and then it can be dropped [here](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/37f1c76e717425ea587650c809f3906228b235e5/app/vminsert/datadogv1/request_handler.go#L69) without any metric decrement

Reply:
also see a comment [here](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/37f1c76e717425ea587650c809f3906228b235e5/app/vminsert/native/request_handler.go#L38)